File: src/com/l2kt/loginserver/SelectorHelper.java
package com.l2kt.loginserver;

import java.nio.channels.SocketChannel;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import com.l2kt.util.IPv4Filter;
import com.l2kt.commons.mmocore.IAcceptFilter;
import com.l2kt.commons.mmocore.IClientFactory;
import com.l2kt.commons.mmocore.IMMOExecutor;
import com.l2kt.commons.mmocore.MMOConnection;
import com.l2kt.commons.mmocore.ReceivablePacket;

import com.l2kt.loginserver.network.LoginClient;
import com.l2kt.loginserver.network.serverpackets.Init;

public class SelectorHelper implements IMMOExecutor<LoginClient>, IClientFactory<LoginClient>, IAcceptFilter
{
	private final ThreadPoolExecutor _generalPacketsThreadPool;
	
	private final IPv4Filter _ipv4filter;
	
	public SelectorHelper()
	{
		_generalPacketsThreadPool = new ThreadPoolExecutor(4, 6, 15L, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>());
		_ipv4filter = new IPv4Filter();
	}
	
	@Override
	public void execute(ReceivablePacket<LoginClient> packet)
	{
		_generalPacketsThreadPool.execute(packet);
	}
	
	@Override
	public LoginClient create(MMOConnection<LoginClient> con)
	{
		LoginClient client = new LoginClient(con);
		client.sendPacket(new Init(client));
		return client;
	}
	
	@Override
	public boolean accept(SocketChannel sc)
	{
		return _ipv4filter.accept(sc) && !LoginController.getInstance().isBannedAddress(sc.socket().getInetAddress());
	}
}

File: src/com/l2kt/loginserver/SelectorHelper.kt
package com.l2kt.loginserver

import com.l2kt.commons.mmocore.*
import com.l2kt.loginserver.network.LoginClient
import com.l2kt.loginserver.network.serverpackets.Init
import com.l2kt.util.IPv4Filter
import java.nio.channels.SocketChannel
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

class SelectorHelper : IMMOExecutor<LoginClient>, IClientFactory<LoginClient>, IAcceptFilter {
    private val _generalPacketsThreadPool: ThreadPoolExecutor =
        ThreadPoolExecutor(4, 6, 15L, TimeUnit.SECONDS, LinkedBlockingQueue())

    private val _ipv4filter: IPv4Filter = IPv4Filter()

    override fun execute(packet: ReceivablePacket<LoginClient>) {
        _generalPacketsThreadPool.execute(packet)
    }

    override fun create(con: MMOConnection<LoginClient>): LoginClient {
        val client = LoginClient(con)
        client.sendPacket(Init(client))
        return client
    }

    override fun accept(sc: SocketChannel): Boolean {
        return _ipv4filter.accept(sc) && !LoginController.isBannedAddress(sc.socket().inetAddress)
    }
}